### PR TITLE
feat: use real pty

### PR DESCRIPTION
**Pull request by caarlos0**
This allows to allocate a real PTY for the session, allowing to do things like running vim within a SSH session.

depends on https://github.com/charmbracelet/bubbletea/pull/879
depends on https://github.com/charmbracelet/ssh/pull/8
closes https://github.com/charmbracelet/wish/issues/196

### CODE REVIEW

**Review by aymanbagabas — COMMENTED** [review]
If we're going to do that, we have to remove the ssh PtyWriter hack, and detect the color profile from the PTY https://github.com/charmbracelet/wish/pull/197/files#diff-79d3f1c4cf631bd3e71a422f4d73666477d1caa5b44e965c98d80e6551f4f8cfR93

**Comment by caarlos0** [comment]
> If we're going to do that, we have to remove the ssh PtyWriter hack, and detect the color profile from the PTY https://github.com/charmbracelet/wish/pull/197/files#diff-79d3f1c4cf631bd3e71a422f4d73666477d1caa5b44e965c98d80e6551f4f8cfR93

Can be done in another PR, as its not strictly related

**Comment by caarlos0** [comment]
still pending changes on ssh cc/ @aymanbagabas

File: bubbletea/tea.go [hunk 1]
```diff
@@ -33,73 +35,81 @@ type Handler func(ssh.Session) (tea.Model, []tea.ProgramOption)
...
+			if !ok {
+				wish.Fatalln(s, "no active terminal, skipping")
+				return
+			}
+
+			upty, ok := tty.Pty.(pty.UnixPty)
+			if ok {
+				f := upty.Slave()
+				out := termenv.NewOutput(f, termenv.WithColorCache(true))
+				opts = []tea.ProgramOption{tea.WithInput(f), tea.WithOutput(out)}
```
Comment (aymanbagabas) [inline bubbletea/tea.go:60]:
```suggestion
				opts = []tea.ProgramOption{tea.WithInput(f), tea.WithOutput(f)}
```

File: bubbletea/tea.go [hunk 2]
```diff
@@ -24,56 +23,34 @@ type BubbleTeaHandler = Handler // nolint: revive
...
-		if m == nil {
-			return nil
-		}
-		opts = append(opts, tea.WithInput(s), tea.WithOutput(s))
-		return tea.NewProgram(m, opts...)
-	}
-	return MiddlewareWithProgramHandler(h, cp)
+// NewRenderer returns a lipgloss renderer for the current session.
+// This function handle PTYs as well, and should be used to style your application.
+func NewRenderer(s ssh.Session) *lipgloss.Renderer {
```
Comment (aymanbagabas) [inline bubbletea/tea.go:28]: Maybe `MakeRenderer` is more suitable since the arguments are fixed

File: bubbletea/tea_other.go [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+func makeIOOpts(s ssh.Session) []tea.ProgramOption {
+	return []tea.ProgramOption{
+		tea.WithInput(s),
+		tea.WithOutput(s),
+	}
+}
+
+func newRenderer(s ssh.Session) *lipgloss.Renderer {
+	return lipgloss.NewRenderer(s, termenv.WithColorCache(true))
```
Comment (aymanbagabas) [inline bubbletea/tea_other.go:21]:
```suggestion
	return lipgloss.NewRenderer(s, termenv.WithUnsafe(), termenv.WithColorCache(true))
```

File: bubbletea/tea_unix.go [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+	return []tea.ProgramOption{
+		tea.WithInput(pty.Slave),
+		tea.WithOutput(pty.Slave),
+	}
+}
+
+func newRenderer(s ssh.Session) *lipgloss.Renderer {
+	pty, _, _ := s.Pty()
+	return lipgloss.NewRenderer(pty.Slave, termenv.WithColorCache(true))
```
Comment (aymanbagabas) [inline bubbletea/tea_unix.go:30]: `pty.Slave` might be nil

**Review by aymanbagabas — APPROVED** [review]

**Comment by caarlos0** [comment]
anyone against this not being a breaking change anymore?

**Comment by aymanbagabas** [comment]
> anyone against this not being a breaking change anymore?

LGTM

**Comment by matthewp** [comment]
@caarlos0 Do you plan on doing a release for this?

**Pull request merged**